lineapy/transformer/linea_transformer.py

```diff
@@ -0,0 +1,171 @@
+import ast
+import logging
+from pathlib import Path
+from types import ModuleType
+from typing import Optional, Union
+
+from lineapy.api.api import _artifact_store, _get, _save, healthcheck
+from lineapy.data.types import (
+    LineaCallNode,
+    Node,
+    PositionalArgument,
+    SourceCode,
+    SourceCodeLocation,
+    SourceLocation,
+)
+from lineapy.exceptions.db_exceptions import ArtifactSaveException
+from lineapy.instrumentation.tracer import Tracer
+from lineapy.utils.utils import get_new_id
+
+logger = logging.getLogger(__name__)
+
+
+class LineaTransformer(ast.NodeTransformer):
+    def __init__(
+        self,
+        code: str,
+        location: SourceCodeLocation,
+        tracer: Tracer,
+    ):
+        self.source_code = SourceCode(
+            id=get_new_id(), code=code, location=location
+        )
+        # tracer.db.write_source_code(self.source_code)
+        self.tracer = tracer
+        # Set __file__ to the pathname of the file
+        if isinstance(location, Path):
+            tracer.executor.module_file = str(location)
+        # The result of the last line, a node if it was an expression,
+        # None if it was a statement. Used by ipython to grab the last value
+        self.last_statement_result: Optional[Node] = None
+
+    def lgeneric_visit(self, node: ast.AST):
+        return node
+
+    def lvisit(self, node):
+        """Visit a node."""
+        method = "lvisit_" + node.__class__.__name__
+        visitor = getattr(self, method, self.lgeneric_visit)
+        return visitor(node)
+
+    def visit_Call(
+        self, node: ast.Call
+    ) -> Union[None, ast.Call, LineaCallNode]:
+        func = self.lvisit(node.func)
+        if func is not None and func[0] == "lineapy":
+            args_to_pass = []
+            to_exec = healthcheck
+            artifact_name: Optional[str] = None
+            argument_values = [self.lvisit(arg) for arg in node.args]
+            dependent = []
+            # keyword_values = {key:value for }
+            if func[1] == "save":
+                artifact_name = argument_values[1]
+                # this is to locate this node in the graph. its predecessors
+                # are the var we are trying to save as an artifact
+                # there can only be one - for now
+                arg = argument_values[0]
+                if isinstance(arg, tuple):
```

Comment:
I think this is risky, what if I do `lineapy.save(foo.bar, "foo bar val")`? Not all `x.y` is our `lineapy.file_system`. What if I do `from lineapy import file_system`? Then it's not even a tuple.